metabase/metabase
# Fix Filter Pill display names for dates

Fixes display labels for weeks in https://github.com/metabase/metabase/issues/12496:

### Description 

Fixes the display names in the Filter Pills for various date-related filters.  DOES NOT fix the date picker for modifying the filter, just fixes the display name so that the filter is accurately represented.

✅ EDIT: All “between” labels in the screenshots below have been replaced with “is” after [design feedback](https://metaboat.slack.com/archives/C02H619CJ8K/p1688777083722309?thread_ts=1688776046.384209&cid=C02H619CJ8K)

#### 🐛 Bug fix: week-related filters

A filter for a single week was being displayed as a day instead of a range of days.  Also, the range of weeks was displayed as an incorrect range of days.

type  | before | after
---|---|---
single week | <img width="217" alt="CleanShot 2023-07-04 at 05 54 10@2x" src="https://github.com/metabase/metabase/assets/116838/9106fa89-525a-4647-8a1a-7ba51341eb1d"> | <img width="239" alt="CleanShot 2023-07-04 at 05 54 19@2x" src="https://github.com/metabase/metabase/assets/116838/d85181c9-c05e-4d3e-b108-6a2b17561224">
week range | <img width="413" alt="CleanShot 2023-07-04 at 06 09 01@2x" src="https://github.com/metabase/metabase/assets/116838/12fd51dd-204d-49ef-a0eb-e2071d09213e"> | <img width="382" alt="CleanShot 2023-07-04 at 06 10 04@2x" src="https://github.com/metabase/metabase/assets/116838/c8b30a63-f072-42b1-95e6-0cd9adf07f2b">

#### 🐛 Bug fix: splice-related filters

Splice filters select a slice of time, spliced across recurrent periods.  

type  | before | after
---|---|---
minute of hour | <img width="143" alt="CleanShot 2023-07-04 at 05 57 43@2x" src="https://github.com/metabase/metabase/assets/116838/85ca0d4d-3759-437c-9d7e-a9edea2a891f">  | <img width="190" alt="CleanShot 2023-07-04 at 05 57 09@2x" src="https://github.com/metabase/metabase/assets/116838/948071c0-fc5c-45fd-979a-28ac2076a5d4">
hour of day | <img width="280" alt="CleanShot 2023-07-04 at 06 00 35@2x" src="https://github.com/metabase/metabase/assets/116838/42534f2c-1e25-443d-8752-23abcc5cd115"> | <img width="179" alt="CleanShot 2023-07-04 at 06 02 29@2x" src="https://github.com/metabase/metabase/assets/116838/23a16e6c-8ace-41f4-b496-4a7ed2611a5f">
day of month | <img width="153" alt="CleanShot 2023-07-04 at 06 04 46@2x" src="https://github.com/metabase/metabase/assets/116838/9697deec-e6d5-4700-999f-16744a2803be"> | <img width="250" alt="CleanShot 2023-07-04 at 06 03 47@2x" src="https://github.com/metabase/metabase/assets/116838/e4f387ad-0eb9-4792-8a0b-afa162a84d78">
day of year | <img width="157" alt="CleanShot 2023-07-04 at 06 05 51@2x" src="https://github.com/metabase/metabase/assets/116838/4a08695b-3d51-47f6-a32e-5f4a75115d1b"> | <img width="244" alt="CleanShot 2023-07-04 at 06 06 38@2x" src="https://github.com/metabase/metabase/assets/116838/6f86504f-c99f-4720-9b98-922777c6db4d">
week of year | <img width="165" alt="CleanShot 2023-07-04 at 06 07 39@2x" src="https://github.com/metabase/metabase/assets/116838/de7cb16f-0cb2-4654-b7db-8280f0203357"> | <img width="245" alt="CleanShot 2023-07-04 at 06 07 02@2x" src="https://github.com/metabase/metabase/assets/116838/65ae8b96-3d66-48b2-99c5-9f7c2dc045cf">

#### 🔧 Improvement: properly formatted date ranges

> Go the extra mile to make the user experience pleasant
> _—The Zen of Metabase_

Dates in a range were originally formatted separately, then joined with a space between.  Since I needed to properly format a date range for the week-filter fix above, I added a general date range formatter based on [Wikipedia’s style guide for date ranges](https://en.wikipedia.org/wiki/Wikipedia:Manual_of_Style/Dates_and_numbers#Ranges).

type | before | after
---|---|---
day | <img width="435" alt="CleanShot 2023-07-04 at 06 31 31@2x" src="https://github.com/metabase/metabase/assets/116838/177bf73d-cc98-411c-933b-5acd253a83dd"> | <img width="330" alt="CleanShot 2023-07-04 at 06 27 49@2x" src="https://github.com/metabase/metabase/assets/116838/26a073d7-727e-4ee4-98fb-e88b1f61f613">
month | <img width="369" alt="CleanShot 2023-07-04 at 06 33 11@2x" src="https://github.com/metabase/metabase/assets/116838/ce25316a-cb71-406c-84be-14964c9a4b2a"> | <img width="331" alt="CleanShot 2023-07-04 at 06 27 11@2x" src="https://github.com/metabase/metabase/assets/116838/75db951d-1066-4c5c-8f0f-d2d0ea13cc3e">
month | <img width="322" alt="CleanShot 2023-07-04 at 06 34 39@2x" src="https://github.com/metabase/metabase/assets/116838/554d0c44-7174-4fa4-bda4-41944d4abb05"> | <img width="327" alt="CleanShot 2023-07-04 at 06 26 40@2x" src="https://github.com/metabase/metabase/assets/116838/1b16e823-3919-4bb8-8172-5385ccb1b78a">


### Implementation Notes

`Filter.displayName` touches a lot of date formatting functions used by other components.  To avoid disrupting the [delicate circuitry](https://github.com/metabase/metabase/assets/116838/67d4991b-7363-4060-b050-ff54a6572269) that other components might rely on, I re-route the display name to use its own internal `Filter.betterDateLabel` for only a specific set of filter types, then fall back to the original display name for others.

### How to verify

See https://github.com/metabase/metabase/issues/12496

Set the unit in the dropdown below to see the different [splice filters](https://github.com/metabase/metabase/assets/116838/9fd6b114-256f-4f74-9580-484fe94c67ca), then drag a range inside the plot.

### Demo

https://github.com/metabase/metabase/assets/116838/f39779d1-297a-40ae-999e-bcf16ed3e053



### Checklist
